### PR TITLE
several fixes to scala.concurrent.util.Duration

### PR DESCRIPTION
**Pull request by rkuhn**
- add test cases (migrated from Akka sources)
- add overflow checking (will throw IllegalArgumentException instead of
  giving wrong results)
- make string parsing more precise when giving >100days in nanoseconds
- make method signatures more precise in retaining FiniteDuration
  throughout calculations
- fix mul/div of infinities by negative number
- add Ordering for Deadline (was accidentally left out earlier)

### CODE REVIEW

**Comment by jsuereth** [comment]
Reviewers?

File: src/library/scala/concurrent/util/Duration.scala [hunk 1]
```diff
@@ -289,12 +321,19 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
     case x: FiniteDuration => toNanos compare x.toNanos
     case _                 => -(other compare this)
   }
+
+  private def add(a: Long, b: Long): Long = {
+    val c = a + b
+    // check if the signs of the top bit of both summands differ from the sum
+    if (((a ^ c) & (b ^ c)) < 0) throw new IllegalArgumentException("integer overflow")
+    else c
```
Comment (retronym) [inline src/library/scala/concurrent/util/Duration.scala]: I find the technique proposed here a little clearer:

https://www.securecoding.cert.org/confluence/display/java/NUM00-J.+Detect+or+prevent+integer+overflow

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: It took me about the same time to think through than the one above, but I can change it if you want. None of these are really dead-obvious ;-)

Comment (retronym) [inline src/library/scala/concurrent/util/Duration.scala]: I suggest to change it and add a comment that links to that CERT page.

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: okay, will do

**Comment by rkuhn** [comment]
PLS REBUILD ALL

File: src/library/scala/concurrent/util/Duration.scala [hunk 2]
```diff
@@ -48,9 +129,12 @@ object Duration {
       case "MinusInf" | "-Inf"        => MinusInf
       case _                          =>
         val unitName = s1.reverse takeWhile (_.isLetter) reverse;
-        def length   = JDouble.parseDouble(s1 dropRight unitName.length)
         timeUnit get unitName match {
-          case Some(unit) => Duration(length, unit)
+          case Some(unit) =>
+            val valueStr = s1 dropRight unitName.length
+            val valueD = JDouble.parseDouble(valueStr)
+            if (valueD <= 9007199254740992d && valueD >= -9007199254740992d) Duration(valueD, unit)
```
Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala]: Magic number is magical. Pull out to a named constant and comment what it is for

File: src/library/scala/concurrent/util/Duration.scala [hunk 3]
```diff
@@ -80,124 +164,217 @@ object Duration {
...
-   * Infinite negative duration: lesser than any other and not equal to any other,
-   * including itself.
+   * Infinite duration: less than any other and not equal to any other
+   * but itself. This value closely corresponds to Double.NegativeInfinity, 
+   * matching its semantics in arithmetic operations.
    */
   val MinusInf: Infinite = new Infinite {
     override def toString = "Duration.MinusInf"
-    def compare(other: Duration) = if (other eq this) 0 else -1
+    def compare(other: Duration) = other match {
```
Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala]: Is this an improvement over old impl?

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: no, there was one more case in the meantime and I didn't revert when that went away, will fix

File: src/library/scala/concurrent/util/Duration.scala (hunk 2 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: will do

File: src/library/scala/concurrent/util/Duration.scala [hunk 4]
```diff
@@ -80,124 +164,217 @@ object Duration {
...
+   *
+   * `d, day, h, hour, min, minute, s, sec, second, ms, milli, millisecond, µs, micro, microsecond, ns, nano, nanosecond`
+   * and their pluralized forms (for every but the first mentioned form of each unit, i.e. no "ds", but "days").
+   */
+  def create(length: Long, unit: String): FiniteDuration   = apply(length, unit)
+  /**
+   * Parse String into Duration.  Format is `"<length><unit>"`, where
+   * whitespace is allowed before, between and after the parts. Infinities are
+   * designated by `"Inf"`, `"PlusInf"`, `"+Inf"` and `"-Inf"` or `"MinusInf"`.
+   * Throws exception if format is not parseable.
```
Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala]: What kind of exception?

File: src/library/scala/concurrent/util/Duration.scala [hunk 5]
```diff
@@ -289,20 +626,56 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
...
+    else if (factor.isNaN) Undefined
+    else if ((factor > 0) ^ (this < Zero)) Inf
+    else MinusInf
+
+  def /(factor: Double) =
+    if (!factor.isInfinite) fromNanos(toNanos / factor)
+    else if (factor.isNaN) Undefined
+    else Zero
+
+  // if this is made a constant, then scalac will elide the conditional and always return +0.0, SI-6331
```
Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala:699]: @retronym said he'd take a stab at SI-6331 (for teh record)

File: src/library/scala/concurrent/util/Duration.scala [hunk 6]
```diff
@@ -80,124 +164,217 @@ object Duration {
...
-      else throw new IllegalArgumentException("illegal subtraction of infinities")
 
-    def *(factor: Double): Duration = this
-    def /(factor: Double): Duration = this
+    def *(factor: Double): Duration =
+      if (factor == 0d || factor.isNaN) Undefined
+      else if (factor < 0d) -this
+      else this
+    def /(factor: Double): Duration =
+      if (factor == 0d || factor.isNaN || factor.isInfinite) Undefined
```
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala]: I'd suggest Inf / 0 => Inf instead of Inf/0 => Undefined, consistently with the behavior of /(other: Duration), and being careful with the sign. I'd try to reuse the tricky code for deciding the sign, that is, `(if ((this > Zero) ^ (other >= Zero)) -1 else 1)`.
UPDATE: On second thought, assuming `-0d < 0d`, it's enough to just drop the first case, `factor == 0d ||`.

File: src/library/scala/concurrent/util/Duration.scala (hunk 4 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: good point

File: src/library/scala/concurrent/util/Duration.scala (hunk 5 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala:699]: which is why I put this comment with the reference in, so that the code can be “normalized” as soon as that issue is fixed

File: src/library/scala/concurrent/util/Duration.scala (hunk 6 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: good point, you found a missing test case; your UPDATE is not enough, though, due to the `-0d` sickness it needs to use `compare` instead of `<`.

BTW: is Double (and Float?) the only example of a violation of the equals/compareTo contract on the JVM which actually ships as part of the platform?

File: src/library/scala/concurrent/util/Duration.scala (hunk 2 above)
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala]: I hope you called it MAGIC_NUMBER, for clarity, and this test becomes:  `if(MAGIC_NUMBER is magic)`

File: src/library/scala/concurrent/util/Duration.scala (hunk 6 above)
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala]: Ouch, thanks for the attention. It's good I made my assumption explicit, and bad I couldn't check myself. I agree it's odd that `Double` (and `Float`, yes) give such results, although clearly documented, but AFAIK the equals contract doesn't require consistent results before and after boxing. To me, it's mostly odd that `-0d == 0d` in Java. I'm confused about Scala, where that same expression can ultimately invoke (depending on boxing) `-0d.equals(0d)`.

File: src/library/scala/concurrent/util/Duration.scala [hunk 7]
```diff
@@ -270,9 +611,30 @@ object FiniteDuration {
...
+       * sorted so that the first cases should be most-used ones, because enum
+       * is checked one after the other.
+       */
+      case NANOSECONDS  ⇒ length != Long.MinValue
+      case MICROSECONDS ⇒ length <= 9223372036854775L && length >= -9223372036854775L
+      case MILLISECONDS ⇒ length <= 9223372036854L && length >= -9223372036854L
+      case SECONDS      ⇒ length <= 9223372036L && length >= -9223372036L
+      case MINUTES      ⇒ length <= 153722867L && length >= -153722867L
+      case HOURS        ⇒ length <= 2562047L && length >= -2562047L
+      case DAYS         ⇒ length <= 106751L && length >= -106751L
```
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala]: Could these magic constants be made non-magic as well?

File: src/library/scala/concurrent/util/Duration.scala [hunk 8]
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
...
+   * Check `System.nanoTime` for your platform.
+   */
   def hasTimeLeft(): Boolean = !isOverdue()
+  /**
+   * Determine whether the deadline lies in the past at the point where this method is called.
+   *
+   * '''''Note that on some systems this operation is costly because it entails a system call.'''''
+   * Check `System.nanoTime` for your platform.
+   */
   def isOverdue(): Boolean = (time.toNanos - System.nanoTime()) < 0
```
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala:64]: Why isn't this just timeLeft < Duration.Zero, which is  the specification? Writing it this way introduces some potential for bugs, now or in the future.
The only difference I see is a few object allocations less. Are there platforms where `nanoTime` is fast enough for this to be relevant (making this a non-system-call has been a goal of optimizations in Windows and Linux, I don't remember how successful)?
Of course the result should be the same, but given the possibility of overflow errors being present or introduced by future changes, there should be (outside the ScalaDoc) a comment which explains why this is equivalent, together with (stress?)-tests checking it. After a quick look, I'd propose:
"This is equivalent to timeLeft < Duration.Zero. Since toNanos is guaranteed to never overflow, and Deadline.time is guaranteed to be a positive Duration, there is no risk of overflow and the results are equivalent."
(Note: I didn't check whether there's any overflow risk for `timeLeft < Duration.Zero`, but in the end that code must work since it just uses the public interface as intended).

Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala:64]: Because timeLeft allocates a new object.

File: src/library/scala/concurrent/util/Duration.scala (hunk 7 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: Given that the expressed limit is clearly stated in the class’ scaladoc comment, I don’t think it is worth calculating them at some other place and giving them names (which would also slow down the creation of every single Duration instance). And they wouldn’t really get any less magic by moving them into the companion to give them names, because then you’d have to search for their definition instead seeing a value which many programmers will intuitively associate with the right thing (granted, I also don’t know the digits by heart for all of them, but they are easily verified).

So, long story short, the best I can think of is adding comments saying how they were calculated.

File: src/library/scala/concurrent/util/Duration.scala [hunk 9]
```diff
@@ -80,124 +168,218 @@ object Duration {
...
+      if (nanos > 0) Inf else MinusInf
+    else if (nanos.isNaN)
+      Undefined
+    else if (nanos > Long.MaxValue || nanos < Long.MinValue)
+      throw new IllegalArgumentException("trying to construct too large duration with " + nanos + "ns")
+    else
+      fromNanos((nanos + 0.5).toLong)
+  }
 
+  private val  µs_per_ns = 1000L
```
Comment (daggerrz) [inline src/library/scala/concurrent/util/Duration.scala]: Making these private[this] would improve the performance of fromNanos. This is s.c.u, after all. :)

Comment (paulp) [inline src/library/scala/concurrent/util/Duration.scala]: Actually what will really improve it is making them final, which will cause the constants to be inlined directly, at which point the access level doesn't matter since the field will never be accessed.

Comment (viktorklang) [inline src/library/scala/concurrent/util/Duration.scala]: What's the reason final isn't implied when private?

Comment (paulp) [inline src/library/scala/concurrent/util/Duration.scala]: Consider a file like this:

```
object Foo {
  private val x = 5
  final val y = 10

  def f1 = () => x
  def f2 = () => y
}
```

f1 returns a closure which returns the value of x.  f2 returns a closure which literally returns "10".  If you recompile this with different values, the closure which returns 10 will still return 10.  The one which accesses x will return the new value of x.

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: thanks for noticing, making them `private final val` has exactly the right effect (i.e. inline the constants and hide them from Java).

**Comment by rkuhn** [comment]
PLS REBUILD ALL

**Comment by rkuhn** [comment]
@jsuereth This does not look like my code was even checked out before it failed?

**Comment by jsuereth** [comment]
PLS REBUILD ALL

**Comment by viktorklang** [comment]
@paulp Thanks for the explanation Paul! Isn't that only an issue when you do hot code reloading though?

**Comment by paulp** [comment]
It's an issue anytime there is separate compilation.  If someone compiles against library version 1.0 and you ship 1.1, code built against 1.0 will use the final constant type vals from 1.0, not those from 1.1.  This may be desirable or undesirable in any given case, but the point is that it's not the same.

**Comment by viktorklang** [comment]
@paulp In the example of the method returning a function returning the constant, that would only be true if the call to the method was replaced with the closure, or?

```
class Foo {
  private final val x = 3
  private val y = 5
  def fX = () => x
  def fY = () => y
}

val f = new Foo
f.fX.apply // call to fX, replacing Foo.class would make this return what the new Foo.class.fX returns?
f.fY.apply
```

**Comment by paulp** [comment]
Maybe this will clarify things:

```
// a.scala
class A {
  private val x0 = 1823905
  def x1 = x0
  final val x2 = 1572398
}
// alt-a.scala
class A {
  private val x0 = 0
  def x1 = x0
  final val x2 = 0
}
// b.scala
class B {
  val a = new A
  val y1 = a.x1
  val y2 = a.x2
  override def toString = "(" + y1 + ", " + y2 + ")"
}

object Test {
  def main(args: Array[String]): Unit = {
    println(new B)
  }
}

// demo.sh
#!/bin/sh
#

scalac a.scala b.scala
scala Test
scalac alt-a.scala
scala Test
```

And:

```
% ./demo.sh 
(1823905, 1572398)
(0, 1572398)
```

**Comment by viktorklang** [comment]
@paulp Yes, that I fully understand. However, the "private final val"-case works as expected (make x0 private final).

**Comment by rkuhn** [comment]
@viktorklang @jsuereth would you please review after my latest changes? Thanks.

File: src/library/scala/concurrent/util/Duration.scala [hunk 10]
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
...
+   * Construct a finite duration from the given length and time unit, where the latter is
+   * looked up in a list of string representation. Valid choices are:
+   *
+   * `d, day, h, hour, min, minute, s, sec, second, ms, milli, millisecond, µs, micro, microsecond, ns, nano, nanosecond`
+   * and their pluralized forms (for every but the first mentioned form of each unit, i.e. no "ds", but "days").
+   */
+  def apply(length: Long, unit: String): FiniteDuration   = new FiniteDuration(length,  Duration.timeUnit(unit))
+
+  // Double stores 52 bits mantissa, but there is an implied '1' in front, making the limit 2^53
+  final val maxPreciseDouble = 9007199254740992d
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala:120]: Type ascription police!

Since this is a public api, let's see:

`final val maxPreciseDouble: Double = 9007199254740992d`

Just in case someone ever drops the 'd' from the end in a crazy fit of refactoring.

File: src/library/scala/concurrent/util/Duration.scala [hunk 11]
```diff
@@ -80,124 +168,218 @@ object Duration {
...
-   * including itself.
+   * Infinite duration: greater than any other (apart from Undefined) and not equal to any other
+   * but itself. This value closely corresponds to Double.PositiveInfinity, 
+   * matching its semantics in arithmetic operations.
    */
   val Inf: Infinite = new Infinite {
     override def toString = "Duration.Inf"
-    def compare(other: Duration) = if (other eq this) 0 else 1
+    def compare(other: Duration) = other match {
+      case x if x eq Undefined => -1
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala]: case `Undefined` => -1  ??

File: src/library/scala/concurrent/util/Duration.scala [hunk 12]
```diff
@@ -80,124 +168,218 @@ object Duration {
...
+      case x if x eq this      => 0
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala]: Note: In Fun Hilarity, you can probably safely do:

```
case `Inf` => 0
```

File: src/library/scala/concurrent/util/Duration.scala [hunk 13]
```diff
@@ -80,124 +168,218 @@ object Duration {
...
   }
 
   /**
-   * Infinite duration: greater than any other and not equal to any other,
-   * including itself.
+   * Infinite duration: greater than any other (apart from Undefined) and not equal to any other
+   * but itself. This value closely corresponds to Double.PositiveInfinity, 
+   * matching its semantics in arithmetic operations.
    */
   val Inf: Infinite = new Infinite {
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala:306]: No final?

File: src/library/scala/concurrent/util/Duration.scala [hunk 14]
```diff
@@ -270,9 +611,29 @@ object FiniteDuration {
...
+ */
 class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
   import Duration._
 
+  require(unit match {
+      /*
+       * enforce the 2^63-1 ns limit, must be pos/neg symmetrical because of unary_-
+       */
+      case NANOSECONDS  ⇒ length != Long.MinValue                                     // max_ns = Long.MaxValue
+      case MICROSECONDS ⇒ length <= 9223372036854775L && length >= -9223372036854775L // max_µs = max_ns / 1000
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala]: Holy magic numbers, batman!   Is there no object we can hide these on as final val?

File: src/library/scala/concurrent/util/Duration.scala [hunk 15]
```diff
@@ -289,20 +650,56 @@ class FiniteDuration(val length: Long, val unit: TimeUnit) extends Duration {
     case x: FiniteDuration => toNanos compare x.toNanos
     case _                 => -(other compare this)
   }
+
+  // see https://www.securecoding.cert.org/confluence/display/java/NUM00-J.+Detect+or+prevent+integer+overflow
+  private def safeAdd(a: Long, b: Long): Long = {
+    if ((b > 0) && (a > Long.MaxValue - b) ||
```
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala:669]:
```
`<joke>` Can we put ZERO in a constant to avoid this magic number? `</joke>`
```

File: src/library/scala/concurrent/util/Duration.scala (hunk 10 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala:120]: nope, this needs to be a constant, and unfortunately constants are not allowed to use the police or any armed forces :-(

File: test/files/jvm/duration-tck.scala [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+import scala.concurrent.util._
+import duration._
+import scala.reflect._
+import java.util.concurrent.TimeUnit._
+
+object Test extends App {
+
+  implicit class Assert(val left: Any) extends AnyVal {
+    import Duration.Undefined
+    def mustBe(right: Any) = right match {
```
Comment (jsuereth) [inline test/files/jvm/duration-tck.scala:15]: Sneaking some scalatest in?

File: test/files/jvm/duration-tck.scala [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+      case r: Double if r.isNaN                        => assert(left.asInstanceOf[Double].isNaN, s"$left was not NaN")
+      case r: Double if r == 0 && r.compareTo(0) == -1 => assert(left == 0 && left.asInstanceOf[Double].compareTo(0) == -1, s"$left was not -0.0")
+      case Undefined                                   => assert(left.asInstanceOf[AnyRef] eq Undefined, s"$left was not Undefined")
+      case _                                           => assert(left == right, s"$left was not equal to $right")
+    }
+  }
+
+  def intercept[T <: Exception : ClassTag](code: => Unit) =
```
Comment (jsuereth) [inline test/files/jvm/duration-tck.scala]: This is a highly useful method.  Note: You can define abstract traits in src/partest that you can extend in these tests.   I'd love to see intercept there.

Also note:  `scala.util.control.Exception.catching` should probably look like this a bit too.

File: src/library/scala/concurrent/util/Duration.scala (hunk 11 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: no, Undefined != Undefined (because it is supposed to behave like Double.NaN)

**Comment by jsuereth** [comment]
These changes are a great improvement!  I have a few minor comments, then LGTM.

File: src/library/scala/concurrent/util/Duration.scala (hunk 13 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala:306]: I’m not aware of any effect this would have; am I wrong?

File: src/library/scala/concurrent/util/Duration.scala (hunk 14 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: wouldn’t help, as discussed earlier: they will stay the same numbers, meaning a comment is the best that can be done, and then I’d prefer to have them in-line so people don’t need to go chasing them

File: test/files/jvm/duration-tck.scala (hunk 16 above)
Comment (rkuhn) [inline test/files/jvm/duration-tck.scala:15]: well, the test was ported from Akka and I didn’t want to change it more than necessary

File: test/files/jvm/duration-tck.scala (hunk 17 above)
Comment (rkuhn) [inline test/files/jvm/duration-tck.scala]: yes, I did that in https://github.com/scala/scala/pull/1279, but since these branches are not based upon each other I cannot use it from here.

File: src/library/scala/concurrent/util/Duration.scala (hunk 14 above)
Comment (jsuereth) [inline src/library/scala/concurrent/util/Duration.scala]: Ah, I missed that.   Sounds ok then..., maybe a // about "Don't touch these unless you are (a) God (b) Roland Kuhn (c) have 3 or more reviewers"

File: test/files/jvm/duration-tck.scala (hunk 17 above)
Comment (jsuereth) [inline test/files/jvm/duration-tck.scala]: Gotcha.  We'll have to do an intercept refactor later.

File: src/library/scala/concurrent/util/Duration.scala (hunk 12 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: actually it turns out that my version produces a lot less byte code, since your proposal will include code handling the case `Inf == null` (as nonsensical as that may seem)

File: src/library/scala/concurrent/util/Duration.scala (hunk 10 above)
Comment (paulp) [inline src/library/scala/concurrent/util/Duration.scala:120]: Yes, I realize I just said "always specify the return type in public api" but there was an implied exception here, because there is no choice.  Anyway, if someone dropped the d it wouldn't even compile.

File: src/library/scala/concurrent/util/Duration.scala (hunk 14 above)
Comment (paulp) [inline src/library/scala/concurrent/util/Duration.scala]: I don't see how a comment is the best which can be done.

```
// a few lines up, it's not much of a chase
final val maxNanos   = Long.MaxValue
final val maxMicros  = maxNanos / 1000
final val maxMillis  = maxMicros / 1000
final val maxSeconds = maxMillis / 1000
final val maxMinutes = maxSeconds / 60
final val maxHours   = maxMinutes / 60
final val maxDays    = maxHours / 24

import FiniteDuration._

require(unit match {
    /*
     * enforce the 2^63-1 ns limit, must be pos/neg symmetrical because of unary_-
     */
    case NANOSECONDS  ⇒   -maxNanos <= length && length <= maxNanos
    case MICROSECONDS ⇒  -maxMicros <= length && length <= maxMicros
    case MILLISECONDS ⇒  -maxMillis <= length && length <= maxMillis
    case SECONDS      ⇒ -maxSeconds <= length && length <= maxSeconds
    case MINUTES      ⇒ -maxMinutes <= length && length <= maxMinutes
    case HOURS        ⇒   -maxHours <= length && length <= maxHours
    case DAYS         ⇒    -maxDays <= length && length <= maxDays
    case _ ⇒
      val v = DAYS.convert(length, unit)
      -maxDays <= v && v <= maxDays
  }, "Duration is limited to +-(2^63-1)ns (ca. 292 years)")
```

You don't have to hardcode long strings of digits.

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: Indeed, it turns out that my worries about the constants turning public were unfounded, and the overhead of one `lneg` in each case surely gets lost in the noise. I’m running the tests again and will push a fix shortly.

**Comment by rkuhn** [comment]
PLS REBUILD ALL

File: test/files/jvm/duration-tck.scala (hunk 17 above)
Comment (Blaisorblade) [inline test/files/jvm/duration-tck.scala]: Notice: I remember at least another use of intercept (and friends) in FutureTests, TryTests, etc.

File: src/library/scala/concurrent/util/Duration.scala [hunk 18]
```diff
@@ -10,36 +10,121 @@ package scala.concurrent.util
...
+ * 3.seconds.fromNow
+ * }}}
+ *
+ * Its main purpose is to manage repeated attempts to achieve something (like
+ * awaiting a condition) by offering the methods `hasTimeLeft` and `timeLeft`.  All
+ * durations are measured according to `System.nanoTime` aka wall-time; this
+ * does not take into account changes to the system clock (such as leap
+ * seconds).
+ */
+case class Deadline private (time: Duration) extends Ordered[Deadline] {
```
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala]: It seems that `time` is in fact required to be a `FiniteDuration`, because e.g. `isOverdue` fails if `time` is not a `FiniteDuration`; if so, this should be checked statically by making `time`'s type `FiniteDuration`. Consequently, you'd have to move `fromNow` into `FiniteDuration` to make it typecheck.
If instead `Inf.fromNow` is supposed to be legal, then `isOverdue` is buggy.

File: src/library/scala/concurrent/util/Duration.scala (hunk 8 above)
Comment (Blaisorblade) [inline src/library/scala/concurrent/util/Duration.scala:64]: Still, the invariant should be documented and tested. And on second thought, there's a bug already:

```
scala> Duration.Inf.fromNow
res0: scala.concurrent.util.Deadline = Deadline(Duration.Inf)

scala> Duration.Inf.fromNow.isOverdue
java.lang.IllegalArgumentException: toNanos not allowed on infinite Durations
    at scala.concurrent.util.Duration$Infinite$class.toNanos(Duration.scala:171)
    at scala.concurrent.util.Duration$$anon$2.toNanos(Duration.scala:186)
    at scala.concurrent.util.Deadline.isOverdue(Duration.scala:22)
(...)
```

Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala:64]: you are right, will fix by requiring FiniteDuration (nothing else makes sense)

File: src/library/scala/concurrent/util/Duration.scala (hunk 18 above)
Comment (rkuhn) [inline src/library/scala/concurrent/util/Duration.scala]: Thanks, I did exactly as you proposed (move fromNow into FiniteDuration) and added a negative test so that `(x: Duration).fromNow` does not compile.

**Comment by rkuhn** [comment]
PLS REBUILD ALL

**Comment by rkuhn** [comment]
@jsuereth since you merged #1279 yesterday, there will be a (logical) conflict now between this and master. Shall I merge master in here and fix it up?

**Comment by rkuhn** [comment]
PLS REBUILD ALL

**Comment by jsuereth** [comment]
Thanks for the hard work and cleanup!
